docs: add VictoriaMetrics architectures guide from startups to hypers… (#9919)
VictoriaMetrics/VictoriaMetrics: docs/guides/vm-architectures/README.md

**Pros:**

* **Disaster Recovery:** The system can survive a complete failure of one cluster's location (AZ or region).  
* **Isolation:** Incidents, maintenance, or configuration errors in one cluster do not affect the others.

**Cons:**

* **Increased Cost:** You are paying more for the infrastructure (compute, storage, and network). The capacity of vmstorage in each AZ is underutilized, since every AZ must be ready to absorb the full traffic load in case of failure. Overhead is \~100% with 2 AZs (50% utilization). For other components, it is possible to use [VPA](https://kubernetes.io/docs/concepts/workloads/autoscaling) or [HPA](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale). 

**Schema:**

![Multi-AZ Architecture](multi-az-architecture.webp)

### Unavailability Scenarios

**Blast radius:** Availability zone

* **Primary region failure (Active-Passive):** switchover in minutes; stale reads until DNS/load balancer/BGP reroute.

* **Single AZ/cluster failure (Active-Active):** seamless reroute; read results may temporarily differ between clusters if cross-AZ replication lags.

* **Cross-region link failure:**

  * Writes: buffered by vmagent.  

>   * Reads: may return stale data until the link is restored.

Reviewer: this can be solved by replicating data on level above.

## The Hyperscale (Cell-based)

**For whom:** For systems that require extra reliability and scalability across multiple regions and zones.

**Key characteristics:** This architecture is built on two main ideas \- cells and the separation of routing and storage paths

> First, we have logical groups of Availability Zones (AZs). Think of these as our data pods. Inside these groups, we deploy our basic clusters. The data within these groups can be either fully replicated (an identical copy in each AZ) or sharded (each AZ holds a distributed portion of the data. For example, in a sharded setup with a replicate data in 3 across 4 cells, each cell will store approximately 75% of all metrics.). 

Reviewer: please, simplify this paragraph. Full sentences in braces are too much :)

Author: solved

---
build:
  list: never
  publishResources: false
  render: never
sitemap:
  disable: true
---

## Matching Architecture to Risk

The complexity of any monitoring system is not an end in itself. It is a direct response to two questions: what risks are we protecting against, and how much performance do we need? This guide is designed to help you choose an architecture that precisely matches your answers.

### Availability as a Guarantee Against Risk

> It's a common mistake to think of availability as a simple number. In reality, availability is a guarantee against a specific level of risk. For example, 99.9% ("three nines") availability allows for about 44 minutes of downtime per month. Before chasing higher nines, ask yourself: is this level of downtime acceptable for your defined risks? Remember that each additional 'nine' of availability often comes with an exponential increase in both system complexity and operational cost.

Reviewer 2: ```suggestion
Availability is commonly mistaken for being a simple metric. In reality, availability is a guarantee against a specific level of risk. For example, 99.9% ("three nines") availability allows for about 44 minutes of downtime per month. Before committing to higher targets, ask yourself: is this level of downtime acceptable for your defined risks? Remember that each additional 'nine' of availability often comes with an exponential increase in both system complexity and operational cost.
```

The scope of the failure you are designing for is your **"blast radius".** Before choosing an architecture, you must first define the blast radius you need to withstand.

### Resilience and Scalability

It is also crucial to distinguish between two fundamental goals:

* **Resilience (or Availability)** is about surviving failures. We achieve it by creating copies ([replicas](https://docs.victoriametrics.com/cluster-victoriametrics/#replication-and-data-safety)) of our components and data.  
* **Scalability (or Performance)** is about handling load. We achieve it by [adding more components](https://docs.victoriametrics.com/victoriametrics/#scalability-and-cluster-version) on every layer to distribute the work.

The architectures in this guide are simply different combinations of these two approaches, designed to handle a specific blast radius.

> ### Architectures as Answers to Blast Radius

Reviewer 2: ```suggestion
### Architecture types as reponse to Blast Radius
```

> Each subsequent section of this guide presents an architecture designed to handle a specific blast radius, moving from the most straightforward setup to the most resilient.

Reviewer 2: ```suggestion
Each subsequent section of this guide presents an architecture designed to contain a specific blast radius, moving from the least complicated setup to the most resilient.
```
move from least to most, complicated -> resilient

> * **[Basic](#basic) (No Resilience).** This architecture is the baseline for non-critical systems. It has no fault tolerance, and its blast radius is the instance itself. Any failure leads to a complete outage.  

Reviewer 2: ```suggestion
* **[Basic](#basic) (No Resilience).** This architecture is the baseline for non-critical systems. This setup has no fault tolerance, and its blast radius is the entire instance. Any failure results in a complete outage.  
```

* **[Single AZ Cluster](#single-availability-zone) (Node-Level Resilience).** This architecture protects against the failure of individual servers (nodes) or application instances within a single Availability Zone. However, its blast radius is the entire AZ; it will not survive a datacenter-wide outage.  
* **[Multi-Cluster and Multi-AZ](#multi-cluster-and-multi-az) (Cluster/AZ/Datacenter-Level Resilience).** Designed as a disaster recovery solution, this architecture can withstand the complete failure of an entire availability zone or data center.  
* **[Hyperscale](#the-hyperscale-cell-based) (Cell/AZ and Region-Level Resilience).** An advanced architecture is built to survive failures of entire Availability Zones or logical "cells" within a region, often degrading gracefully instead of failing.  
* **[Logical Layers](#logical-layers) (Logical Resilience).** This is not a physical resilience level but an architectural layer on top of any setup. It addresses the risk of data access conflicts by providing strong logical isolation between different teams or customers.

> ### The decision tree

Reviewer 2: ```suggestion
### The Decision Tree
```
Camel-cased to be consistent across other headlines

<p align="center">
<img src="decision-tree.webp" alt="Decision Tree" width="80%">
</p>

## Basic

**For whom**: pet projects, developments/test stages, and non-critical systems monitoring.

Installation guide reference: [https://docs.victoriametrics.com/guides/k8s-monitoring-via-vm-single](https://docs.victoriametrics.com/guides/k8s-monitoring-via-vm-single)

> **Key characteristics**: Single instance that does everything: stores, retrieves, and provides metrics.

Reviewer 2: ```suggestion
**Key characteristics**: Single instance that does everything: stores, retrieves, and provides metrics, minimal footprint
```

**Pros**:

* **Straightforward.** Quick deployment without additional components  

> * **Cost-efficient.** It avoids redundant work, such as writing or transmitting the same data twice, thereby reducing both computational and network expenses. Additionally, there are no extra copies of data.

Reviewer 2: ```suggestion
* **Cost-efficiency.** It avoids redundancy, such as writing, transmitting, or storing the same data twice, thereby reducing both computational and network expenses.
```

**Cons**:

* **Single point of failure.** No fault tolerance and no availability

**Schema:**

<p align="center">
<img src="basic-architecture.webp" alt="Basic Architecture" width="40%">
</p>

### Unavailability Scenarios

In this simplest setup, any single-node failure leads to temporary data unavailability or loss until the instance restarts or storage is restored. There are no built-in redundancy or replication layers.

For this section, you can increase availability by utilizing backup and restore mechanisms on various levels: hardware, virtualization, persistence volume management, or application. VictoriaMetrics provides the [backup tools](https://docs.victoriametrics.com/victoriametrics/vmbackup/) to achieve that. 

## Single Availability Zone

> **Target audience**: Single availability zone hosted system of any scale

Reviewer 2: ```suggestion
**Target audience**: System of any scale hosted in a single availability zone 
```

Installation guide reference: [https://docs.victoriametrics.com/guides/k8s-monitoring-via-vm-cluster](https://docs.victoriametrics.com/guides/k8s-monitoring-via-vm-cluster)

High availability implementation: [https://docs.victoriametrics.com/guides/k8s-ha-monitoring-via-vm-cluster](https://docs.victoriametrics.com/guides/k8s-ha-monitoring-via-vm-cluster)

> **Key characteristics:** This is a complete VictoriaMetrics cluster, commonly running in a single Kubernetes cluster. Each component of the cluster: vminsert, vmselect, and vmstorage has multiple copies (replicas). The data is also copied and sharded between vmstorage nodes using the `--replicationFactor` setting on vminsert. [See the official documentation](https://docs.victoriametrics.com/victoriametrics/cluster-victoriametrics/#replication-and-data-safety) to determine the optimal replication factor for your needs.

Reviewer 2: ```suggestion
**Key characteristics:** This is a full VictoriaMetrics cluster, commonly running in a single Kubernetes cluster. Each component of the cluster: vminsert, vmselect, and vmstorage has multiple copies (replicas). The data is also copied and sharded between vmstorage nodes using the `--replicationFactor` setting on vminsert. [See the official documentation](https://docs.victoriametrics.com/victoriametrics/cluster-victoriametrics/#replication-and-data-safety) to determine the optimal replication factor for your needs.
```
Not sure if its "copied and sharded" - we replicate it, but don't shard it, don't we?

**Pros**:

* **Reliability.** The system can survive a failure of any component without service disruption. If a single vmstorage node dies, other replicas continue to operate. And it is the same for other components.

**Cons**:

* **No disaster recovery.** If the entire Kubernetes cluster, availability zone, or data center fails, you lose the entire monitoring system.  

> * **Increased Cost:** Storage cost grows linearly with the replicationFactor (e.g., RF=2 equals 2x storage, RF=3 equals 3x).   Compute components like vminsert or vmselect scale horizontally and increase throughput rather than duplicating data.

Reviewer 2: ```suggestion
* **Increased Cost:** Storage cost grows linearly with the replicationFactor (e.g., RF=2 equals 2x storage, RF=3 equals 3x). Compute components like vminsert or vmselect scale horizontally and increase throughput rather than duplicate data.
```

**Schema:**

<p align="center">
<img src="single-az-architecture.webp" alt="Single AZ Architecture" width="60%">
</p>

### Application vs. Storage Replication

When building a resilient cluster, several replication options are available.

> **Path A: Application-Level Replication.** This approach is enabled [by setting](https://docs.victoriametrics.com/victoriametrics/cluster-victoriametrics/#replication-and-data-safety) the `-replicationFactor=N` flag, where N is an integer representing the desired number of replicas. It makes the cluster components responsible for writing N copies of the data across different vmstorage nodes.

Reviewer 2: ```suggestion
**Path A: Application-Level Replication.** This approach is enabled by [`-replicationFactor=N` flag](https://docs.victoriametrics.com/victoriametrics/cluster-victoriametrics/#replication-and-data-safety), where N represents desired number of replicas. It makes the cluster components responsible for writing N copies of the data across different vmstorage nodes.
```

Reviewer 2: Also needs "best for"/"target audience" section and "key characteristics"

**Pros:**

* **Guaranteed Query Completeness on Node Failure.** The key advantage is that the cluster is aware of its replication. It can survive a complete vmstorage node failure and still guarantee 100% complete query responses from the remaining replicas (as long as the number of failed nodes is less than the replication factor).  

> * **Infrastructure-Independent Logic.** The replication logic is part of the VictoriaMetrics application, ensuring the same predictable behavior whether you run on-premise or on any cloud provider.

Reviewer 2: ```suggestion
* **Infrastructure-Independent Logic.** The replication logic is part of the VictoriaMetrics application, ensuring the same predictable behavior regardless of the underlying infrastructure.
```

**Cons:**

* **Latency sensitivity risks**. A slow or overloaded replica can increase write latency, since inserts must complete on multiple nodes. A larger number of nodes increases the risk of problems with one of them.

**Path B: Storage-Level Replication (The Cloud Provider Way)** In this model, VictoriaMetrics replication factor is set to 1, and the vmstorage data is backed up with cloud-provided and replicated volumes(i.e., AWS EBS replicated within AZ, Google Zonal PD). 

**Pros:**

* **Offloaded Resource Cost.** The data replication is no longer bound by application CPU and network performance, and is offloaded to the cloud provider's storage infrastructure. 

**Cons:**

> * **No read resilience.** Any vmstorage restart (including planned maintenance) or failure makes its data temporarily unavailable for querying.

Reviewer 2: ```suggestion
* **No read resilience.** Any planned or unplanned unavailability (e.g., failure or maintenance) on vmstorage node renders its data temporarily unavailable for querying.
```

Reviewer 2: We should address "maintenance" here - there should be a method to perform it

* **Failover duration.** When a node or disk fails, the PVC must be reattached to another node. For zonal volumes (single AZ), this can take seconds to minutes (e.g., 10-60 seconds for clean detach/attach; up to 5 minutes in force-detach cases), making data from that shard temporarily unavailable for querying until reschedule completes.

> ### Query Consistency Partial vs. Complete Responses

Reviewer 2: ```suggestion
### Query Consistency: Partial vs. Complete Responses
```

In a large, distributed system, partial failures are a common occurrence. A critical choice is how your read path should behave when only partial data can be retrieved.

> **Path A: Allow Partial Responses (Focus on Availability)** By default, if a vmstorage node is down, vmselect will continue getting results from the healthy vmstorage nodes. If more than or equal to the replicationFactor vmstorage nodes fail to respond, the response will have the "isPartial" field set to true.

Reviewer 2: ```suggestion
**Path A: Allow Partial Responses (Read availability focus)** When a vmstorage node is down, vmselect will continue getting results from the healthy vmstorage nodes. If no more than N=replicationFactor vmstorage nodes fail to respond, the response will have the "isPartial" field set to true.
```
Not sure what "by default" means here

Needs best for / key advantages section.

**Pros:** 

> * **High Availability.** Some data, albeit incomplete, would still be available. The system degrades gracefully. It will continue to return available data from the remaining healthy nodes, rather than failing the entire query.

Reviewer 2: ```suggestion
* **High Availability.** Some data, albeit incomplete, would still be available. The system degrades gracefully. Available data from the remaining healthy nodes is returned, rather than failing the entire query.
```

**Cons:**

* **Risk of incomplete data.** Users might not realize the "partial" warning and make decisions based on incomplete and possibly misleading graphs.

> **Path B:** **Deny Partial Responses (Focus on Consistency)** You can configure vmselect with the `-search.denyPartialResponse` flag. If vmselect cannot fetch a complete result from all vmstorage nodes that hold the requested data according to the replication factor value, it will return an error instead of a partial result.

Reviewer 2: ```suggestion
**Path B:** **Deny Partial Responses (Consistency focus)** You can configure vmselect with the `-search.denyPartialResponse` flag. If vmselect cannot fetch a complete result from all vmstorage nodes that hold the requested data according to the replication factor value, it will return an error instead of a partial result.
```

best for / key advantages

**Pros:**

* **Guaranteed data consistency.** This approach ensures that any successful query returns 100% of the requested data. If vmselect receives only a partial response from its vmstorage nodes, the entire query is marked as failed, preventing any misleading or incomplete results.

**Cons:**

* **Lower Availability.** This approach sacrifices availability to guarantee consistency. So if more than replicationFactor vmstorage nodes are unavailable, read queries will start returning errors.

### Buffering Strategy Trade-off

Once you have a vmagent sending data to the storage component (vmsingle or cluster), you face your first important trade-off: what should vmagent do when the storage is temporarily unavailable? This choice defines the trade-off between higher availability (by not losing data) and lower resource consumption (by not using disk). By default, vmagent acts as a durable queue: it persists compressed unsent data to the local filesystem. The size of the queue is controlled via \`--remoteWrite.maxDiskUsagePerURL\` and can be [estimated in advance](https://docs.victoriametrics.com/victoriametrics/vmagent/#calculating-disk-space-for-persistence-queue).

> **Path A: Stateful Mode (Most Reliable).**  By default, [the operator uses ephemeral storage](https://docs.victoriametrics.com/operator/resources/vmagent/#statefulmode) for the vmagent queue. In production, we recommend explicitly configuring a PersistentVolumeClaim (PVC) for vmagent to ensure the buffer is stored on a persistent disk and survives pod restarts. [The documentation](https://docs.victoriametrics.com/victoriametrics/vmagent/#on-disk-persistence) about on-disk persistence.

Reviewer 2: ```suggestion
**Path A: Stateful Mode (Most Reliable).**  By default, the operator sets up vmagent to use [ephemeral storage](https://docs.victoriametrics.com/operator/resources/vmagent/#statefulmode) for the vmagent queue. In production, we recommend explicitly configuring a PersistentVolume (PV) for vmagent to ensure the buffer is stored on a persistent disk and survives pod restarts. See docs on [on-disk persistence](https://docs.victoriametrics.com/victoriametrics/vmagent/#on-disk-persistence).
```

Best for / key advantages (persisted data). PVC is a mechanism to request a volume, data is not stored there

**Pros:**

* **Improved reliability.** Unsent data is safe during vmagent restarts or when remote storage is down (until queue is full).

**Cons:**	

* **Requires additional resources.** Deployment becomes stateful, uses disk space, and I/O. The queue size can build extra pressure on remote storage once it becomes available.

> For Enterprise users, the queueing can be offloaded to an external message broker, such as **Kafka**. In that case vmagent can [read or write into Kafka](https://docs.victoriametrics.com/victoriametrics/integrations/kafka/).

Reviewer 2: Shouldn't this be a separate Path?

> **Path B: Ephemeral Buffering (with tmpfs).** For maximum performance, the vmagent buffer directory can be mounted as a tmpfs volume, which is physically stored in the node's RAM. In Kubernetes, this is configured via `emptyDir: { medium: "Memory" }`.

Reviewer 2: Best for / key advantages

**Pros:**

* **Fast I/O.** Buffering happens at RAM speed. This path safeguards against brief network outages without any loss of performance.

**Cons:**

> * **Significant risk of data loss.** Unsent data is lost on vmagent restarts. The queue size is limited by the available memory.

Reviewer 2: ```suggestion
* **increased risk of data loss.** Unsent data is lost on vmagent restarts. The queue size is limited by the available memory.
```
I don't think we should categorize risk, but we should mention the increase in probability

### Unavailability Scenarios

**Blast radius:** Cluster

* **Instance/pod failure:**  

>   * Path A (Application-level replication, RF ≥2): no impact; cluster continues with remaining replicas.  

Reviewer 2: ```suggestion
  * Path A (Application-level replication, RF ≥2): no impact until the number of failed instances is less than the replication factor.  
```

High availability implementation: [https://docs.victoriametrics.com/guides/multi-regional-setup-dedicated-regions](https://docs.victoriametrics.com/guides/multi-regional-setup-dedicated-regions)

**Key characteristics:** The core principle of this architecture is to run two or more independent, self-contained VictoriaMetrics clusters (from the [Single AZ](#single-availability-zone) section) in separate failure domains, such as different Availability Zones or geographic regions. A global, stateless layer is responsible for routing write and read traffic to these clusters. Each participating AZ must be provisioned to handle the entire workload if another AZ fails. 

There are no differences in the VictoriaMetrics clusters' topology regarding the multi-AZ approach. It can be Active-Active or Active-Passive - the schema will be the same.  

To ensure reliability, vmagent implements the bulkhead pattern: each destination URL configured via `--remoteWrite.url` is assigned a dedicated data queue and an isolated pool of workers. This isolates the data streams, ensuring that if one storage destination becomes slow or unavailable, it does not impact data delivery to the others.

**Pros:**

> * **Disaster Recovery:** The system can survive a complete failure of one cluster's location (AZ or region).  

Reviewer 2: ```suggestion
* **Disaster Recovery:** The system can seamlessly survive a complete failure of one location (AZ or region).  
```

**Schema:**

<p align="center">
<img src="multi-az-architecture.webp" alt="Multi-AZ Architecture" width="65%">
</p>

### Unavailability Scenarios

**Blast radius:** Availability zone

* **Primary region failure (Active-Passive):** switchover in minutes; stale reads until DNS/load balancer/BGP reroute.

* **Single AZ/cluster failure (Active-Active):** seamless reroute; read results may temporarily differ between clusters if cross-AZ replication lags.

* **Cross-region link failure:**

  * Writes: buffered by vmagent.  
  * Reads: may return stale data until the link is restored.

## The Hyperscale (Cell-based)

**For whom:** For systems that require extra reliability and scalability across multiple regions and zones.

**Key characteristics:** This architecture is built on two main ideas - cells and the separation of routing and storage paths

> First, we have logical groups of Availability Zones (AZs). Think of these as our data pods. Inside these groups, we deploy our basic clusters. The data within these groups can be distributed in two ways:

Reviewer 2: ```suggestion
First, Availability Zones (AZs) are logically grouped - think of these as our data pods. Inside these groups, we deploy our basic clusters. The data within these groups can be distributed in two ways:
```

We've already defined an abbreviation for AZs (or really should have first)

Reviewer 2: We need to introduce cells earlier - ideally immediately after they were mentioned first

> - **Fully replicated:** An identical copy exists in each AZ.

Reviewer 2: ```suggestion
- **Fully replicated:** An identical cluster replica exists in each AZ.
```
"copy" here means entire infra and data, right?

- **Sharded:** Each AZ holds a portion of the data. For example, with replication factor 3 across 4 cells, each cell stores approximately 75% of all metrics.

> Inside each Storage Cell, the VictoriaMetrics cluster is configured with a `-replicationFactor` of 1. High availability is achieved by replicating data across multiple cells by the global routing layer, not within the cell or the cluster.

Reviewer 2: we mentioned replication factor 3 previously?

> Next, we have a separate, stateless layer of routing cells. Their only purpose is to manage traffic. They accept all incoming data and queries and intelligently route them to the correct storage groups. This separation of routing and storage is key to the design. 

Reviewer 2: ```suggestion
Next, we have a separate, stateless layer of routing cells. Their only purpose is to manage traffic. They accept all incoming data and queries and intelligently route them to the correct storage groups. This separation of routing from storage is key to this design. 
```

### Path A: Prioritize Data Completeness (The Global vmselect model)

In this model, your primary goal is to obtain as complete and consistent data as possible for every query, even if some storage cells are lagging behind.

**Write Path:** vmagent [shards data](https://docs.victoriametrics.com/victoriametrics/vmagent/#sharding-among-remote-storages) across your storage cells. Fault tolerance is configured via `-remoteWrite.shardByURL` and `-remoteWrite.shardByURLReplicas` (for example, writing each time series to 3 out of 4 cells). Redundancy is achieved across cells, not within a cell. This provides resilience against cell failures while saving storage compared to full copies.

**Read Path:** You use a two-level vmselect system. A global vmselect receives user queries. In turn, it queries local vmselects in each of your storage cells and merges the results. Exposing local VMSelects to a global one is necessary because there can be no possibility to connect directly to vmstorage on the local cell, especially if it is in Kubernetes, as there is no HTTP endpoint for querying vmstorage. And using NodePort may not be a good practice for production.

> **Schema:**

Reviewer 2: Move below pros/cons to maintain the same structure

Global vmauth -> Cell -> vmselect

**Pros:**

* **Very fast queries.** There is no overhead from merging results from multiple cells.  
* **Low cross-cell traffic for reads.** This can significantly reduce network costs.

**Cons:**

* **High storage cost.** You are storing redundant, full copies of data, which is an expensive approach.  
* **The Freshness Trap.** This is the greatest and most significant risk associated with this approach. If the write path to one storage cell slows down, vmagent will start buffering data for it. Internally, vmagent maintains a separate queue for each `-remoteWrite.url` target, so lag in a single cell can cause it to serve stale results under the `first_available` policy.  If vmauth sends a user to this cell while its queue is not empty, that user will receive stale data (data that is not 100% fresh). A certain automation could be used to disable reads from cells that are lagging behind.

### Alerting Strategy Trade-offs

Just like the read path, your alerting strategy in a hyperscale setup also involves critical trade-offs.

**Path A: Local vmalert (Fast Evaluation, High Traffic). In this model, you deploy vmalert inside each storage cell.**

> **How it works:** Each vmalert queries its local vmselect for data. This is very fast and efficient. It then sends its firing alerts to a global Alertmanager cluster, which is likely located in the compute cells.

Reviewer 2: Reformat into "best for" / "key advantages" for consistency

**Pros:**

> * **Low latency for alert evaluation.** Queries are always local and fast.

Reviewer 2: ```suggestion
* **Low latency for alert evaluation.** Query evaluation is always local and fast.
```
